gammapy/cube/fit.py

```diff
@@ -172,6 +178,113 @@ def likelihood(self):
 
         return stat
 
+    def to_hdulist(self):
```

Comment:
I think the extension names should be optional arguments here. This is the only way to get some flexibility and backward compatibility.

Reply:
How do you deal with `bands` HDUs? Shouldn't they have their own names? Counts, background and exposure don't have the same bands.

Reply:
The band HDUs are already handled correctly. They are separated by a corresponding prefix in the HDU name. Here is an example HDU list, resulting from a call to `.to_hdulist()`:

```
Filename: (No file associated with this HDUList)
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU       4   ()      
  1  COUNTS        1 ImageHDU        28   (100, 100, 2)   float64   
  2  COUNTS_BANDS    1 BinTableHDU     22   2R x 4C   ['I', 'E', 'E', 'E']   
  3  EXPOSURE      1 ImageHDU        28   (100, 100, 3)   float64   
  4  EXPOSURE_BANDS    1 BinTableHDU     22   3R x 4C   ['I', 'E', 'E', 'E']   
  5  BACKGROUND    1 ImageHDU        28   (100, 100, 2)   float64   
  6  BACKGROUND_BANDS    1 BinTableHDU     22   2R x 4C   ['I', 'E', 'E', 'E']   
  7  EDISP         1 BinTableHDU     31   3R x 6C   ['E', 'E', 'I', 'PI()', 'PI()', 'PE()']   
  8  EDISP_EBOUNDS    1 BinTableHDU     26   2R x 3C   ['I', 'D', 'D']   
  9  PSF           1 ImageHDU        28   (51, 51, 3)   float32   
 10  PSF_BANDS     1 BinTableHDU     22   3R x 4C   ['I', 'E', 'E', 'E']   
 11  MASK_FIT      1 ImageHDU        28   (100, 100, 2)   int64   
 12  MASK_FIT_BANDS    1 BinTableHDU     22   2R x 4C   ['I', 'E', 'E', 'E']   
```

I would be hesitant to introduce the flexibility to let users choose the HDU name. I'd prefer convention over configuration here, until we find a real short-coming of the current naming scheme. I agree that it's useful for backwards compatibility, but I guess we don't want to allow that HDU names are chosen freely either? Otherwise will end up with a lot of incompatible FITS file, where HDUs have to be assigned everytime by hand, because somebody changed the name from `COUNTS` to `ON` or `EDISP` to `ENERGY_DISPERSION`. I'll check how we handle this elsewhere in Gammapy and comment again.

Reply:
> I think the extension names should be optional arguments here. This is the only way to get some flexibility and backward compatibility.

Similar to what @adonath replied, I would also *not* introduce that flexibility, because the more diversity we have, the harder it will be to support old files.

My suggestion would be to merge as-is (with an assert on current HDU names added to make it clear what we have), and maybe, if you want, open up a follow-up issue where we can discuss options how to possibly add versioning or a level of indirection (like a YAML file, or a dict stored as a JSON blob in the primary header) that would allow us to evolve the content and names in the future, while keeping support for older formats. (Not immediately, but after v1.0, for the next few months changes are allowed)